README.md
# S.I.M.O.N.

Code completion is one of the main features of modern Integrated Development Environments (IDEs). Its aim is to assist and facilitate the developer in code writing, by predicting the next code token(s) the developer is likely to write, therefore speeding up the development process. Currently, there are no tools that can be easily adapted to support different prediction models, therefore allowing researchers and practitioners to intuitively and easily experiment with different approaches, until now... This plugin allows experimenters to test their own Java code prediction neural networks in real world scenarios. The functioning of this plugin is very intuitive for the final user. The proposed neural network must be uploaded on a server that respects the simple API of the plugin. In the plugin settings, two addresses can be specified, one for the code completion neural network and one for the comment completion neural network. Once an address has been specified, it is sufficient to normally start programming and recommendations will be displayed in a non-intrusive light grey color directly in the text-editor. At this point, if the recommendation is not relevant to the context, it is sufficient to keep writing to make the suggestion gently fade away. If, on the other hand, the suggestion is relevant to the context, the user can manually write it down or accept it with a keyboard shortcut. The plugin will take care of the business of parsing the java file, understanding the position of the caret (code or comment), triggering recommendations in the correct situations and understanding when recommendations are not needed anymore
Programming is a huge discipline and it can be carried out in different ways. Sometimes plain java code is sufficient to solve a problem, but most of the times the use of external libraries and frameworks is required. Solving such a complex task as token(s) prediction requires lots of data, since a neural network, in order to be effective, must be trained in the largest amount of scenarios. Collecting huge amount of data that spans the greatest amount of use cases is one of the most challenging aspects of building such a system. For this reason, there is a feedback storing feature integrated in the plugin. An additional address can be specified in the settings. This address must lead to a server that respects the simple API of the plugin. Whenever a user accepts or rejects a suggestion, a feedback will be sent to the server. By doing so, it is possible to collect several additional data points in order to further train the network and therefore enhance its capabilities.
In order to experience the beauty of intelligent code prediction, there is no need to build your own neural network. By default the system uses an implementation of the T5 model, specifically tuned and trained by the USI Software Institute to perform token(s) prediction. T5 models represent a viable solution for code completion, outperforming even the powerful RoBERTa models, with perfect predictions ranging from ~30%, obtained when asking the model to guess entire blocks, up to ~69%, reached in the simpler scenario of few tokens masked from the same code statement.

## Features
This plugin uses a very powerful implementation of a T5 plugin to perform code and comment completion. The suggestions coming from the neural network will directly appear in the text-editor.
If you are not satisfied with the predictions coming from my model, you can easily deploy your own model on a server and specify the url in the plugin settings.

![A usage scenario of the plugin](./images/example.gif)


## Extension Settings
This extension contributes the following settings:

* `simon.enable`: enable/disable this extension
* `simon.urlCode`: the server hosting the neural network that makes code recommendations <br>
Default: `http://gym.si.usi.ch:45001/code/`
* `simon.urlComment`: the server hosting the neural network that makes comment recommendations <br>
Default: `http://gym.si.usi.ch:45001/comment/`
* `simon.triggerChars`: specifies the characters for which the autocompletion will be triggered <br>
Default: ` " ", ".", ",", "(" `
* `simon.name`: insert your organization name when requesting a recommendation (optional)
* `simon.urlFeedback`: the server hosting the feedback database <br>
Default: `http://gym.si.usi.ch:45002/feedback/`
* `simon.suggestionColor`: the color of the suggestion is hexadecimal <br>
Default: `#505050` <br>
I suggest to keep this subtle
* `simon.confidenceCode`: the minimum confidence that the code neural network must have in order for a recommendation to be shown <br>
Default: `75`
* `simon.confidenceComment`: the minimum confidence that comment neural network must have in order for a recommendation to be shown <br>
Default: `75`
* `simon.singleLineComment`: enable/disable to move a single line comment on the same line of some code, to a new line when the code is sent to the neural network <br>
Default: `True`

## Commands
These are the default commands of the extension. The following keys change behavior depending on the context. This means that it will still be possible to indent code using `tab`, only when there is a recommendations showing, this key will change its behavior. This same reasoning can be applied to all the other shortcuts.
* `escape`: make the showing suggestion fade away
* `tab`: accept a suggestion
* `ctrl+space`: manually invoke the neural network

## Folder structure

- **`package.json`** - this is the manifest file in which you declare your extension and command. The plugin registers a command and defines its title and command name. With this information VS Code can show the command in the command palette.
- **`src/`**: this folder contains all the javascript code to make the plugin work
- **`language-configuration.json`**: file that disables automatic parenthesis closing. Necessary for the correct operation of the plugin
- **`report.pdf`**: pdf file containing a very detailed description of the project

-----------------------------------------------------------------------------------------------------------

**Note:** This plugin will turn-off automatic closing of (, [ and {


CHANGELOG.md
# Change Log

All notable changes to the "java-autocompletion" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/editor.js
// -------------------------------------------------------------------------------------------------
// this file contains the functions to show, delete and accept the recommendation plus some 
// helper functions
// -------------------------------------------------------------------------------------------------
const vscode = require('vscode');
const neuralNetwork = require('./neuralNetwork')
const handle = require('./handleKeyPresses')
// get the theme of the current vs-code window
const workbenchConfig = vscode.workspace.getConfiguration('workbench')
const theme = workbenchConfig.get('colorTheme')
var editor = vscode.window.activeTextEditor;


// this is the decoration that will be used to show the suggestions
const annotationDecoration = vscode.window.createTextEditorDecorationType({
    after: {
        margin: '0 0 0 0em',
        textDecoration: 'none',
    },
    rangeBehavior: vscode.DecorationRangeBehavior.ClosedOpen,
});

// flag to check if there is a recommendation showing already
// this flag is checked and manipulated by the methods:
// - showSuggestion()
// - removeSuggestion()
// - acceptSuggestion()
var isPresent = false;

// this varaible is manipultaed by 'acceptSuggestion()'
// it is set to true by 'editor/acceptSuggestion()' and 'editor/removeNewLines()' and set to false one second after 
// 'acceptSuggestion()' has terminated. By doing this, 'onDidChangeTextDocument'
// is not going to be called when the suggestion is inserted.
// As a matter of fact, 'onDidChangeTextDocument' is not called only when 
// the user types something but every time the docuemnt is modified.
var isBlocked = false;

// this array of string, will contain the suggestion coming from the NN
// splitted at '\n'
suggestion = [];

// this string holds the suugestion not splitted in an array
s = "";

// this counter represents the number of lines we need to insert
// basically it is equal to suggestion.length - 1
// we subtract 1 from it because if suggestion.length is equal to 1, it means that
// we don't need to add any line, the suggestion will simply show in the same line
// the user is typing in
lines = 0;

// the content of the current method
// this varibale is made public because it is used by extension.js and handleKeyPresses.js
// to send feedback to the server
methodContent = ""


const getIsPresent = () => {
    return isPresent;
}
exports.getIsPresent = getIsPresent;

const setIsPresent = (bool) => {
    isPresent = false;
}
exports.setIsPresent = setIsPresent;

const getIsBlocked = () => {
    return isBlocked;
}
exports.getIsBlocked = getIsBlocked;

const getMethodContent = () => {
    return methodContent
}
exports.getMethodContent = getMethodContent;


/**
 *  insert 'lines' new lines to make room for the suggestion
 * @param {Int} lines the number of lines that must be inserted
 * @returns {Promise}
 */
function insertNewLines(lines) {

    // if lines == 0, there is no need to remove any line
    if (lines == 0) {
        return;
    }

    // create 'lines' new lines
    ins = "";
    for (i = 0; i < lines; i++) {
        ins += "\n"
    }
    editor = vscode.window.activeTextEditor;
    // compute the position in which we need to insert the new lines
    cursor = editor.selection.active;
    startPos = new vscode.Position(cursor.line, cursor.character + 1)
    endPos = new vscode.Position(cursor.line, cursor.character + 2);


    return vscode.window.showTextDocument(editor.document, 1, false).then(e => {
        return e.edit(edit => {
            // insert the lines in the editor
            return edit.replace(new vscode.Range(startPos, endPos), ins)
        })
    }).then(() => {
        // move the caret back to the previous position
        tmpPos = new vscode.Position(cursor.line, cursor.character)
        range = new vscode.Range(tmpPos, tmpPos);
        editor.selection = new vscode.Selection(tmpPos, tmpPos);
        editor.revealRange(range);
    })
}

/**
 * remove the lines that we previosuly inserted using the 'insertNewLine()' method,
 * @param {String} lines the number of lines that must removed
 * @returns {Promise}
 */
function removeNewLines(lines) {
    // if we didn't insert any line, then we don't need to do anything
    if (lines == 0) {
        return;
    }

    editor = vscode.window.activeTextEditor;
    cursor = editor.selection.active;
    // compute the portion of the text we need to remove
    startPos = new vscode.Position(cursor.line, cursor.character + 1)
    endPos = new vscode.Position(cursor.line + lines, 0);

    isBlocked = true;
    return vscode.window.showTextDocument(editor.document, 1, false).then(e => {
        return e.edit(edit => {
            edit.replace(new vscode.Range(startPos, endPos), "")
        })
    }).then(() => {
        isBlocked = false;
    })
}


/**
 * Remove the suggestion from the editor this means to simply remove the decoration using an empty array
 * and to remove the new lines.
 * This method is called when the user doesn't accept the suggestion (for example by keep typing), therefore 
 * the suggestion must disappear as if it had never been there
 * @returns {void}
 */
const removeSuggestion = async () => {
    // if there isn't a recommendation don't even try to delete
    if (isPresent == false) {
        return;
    }

    editor = vscode.window.activeTextEditor;

    // remove the decoration
    editor.setDecorations(annotationDecoration, []);
    // remove the lines that we previously inserted
    await removeNewLines(lines);
    // the recomndation has been deleted, therefore there isn't a recommendation anymore
    isPresent = false;
    // change the behavior of the shortcuts
    vscode.commands.executeCommand('setContext', 'myContext', isPresent);
}

exports.removeSuggestion = removeSuggestion;

/**
 * accept the suggestion, remove the annotation and inject actual code in the editor
 * @param {*} line the line in which the suggestion is shown
 * @param {*} startTyping the position in which the user has just typed.
 * @returns {void}
 */
const acceptSuggestion = async (line, startTyping) => {
    // if there isn't a recommendation, don't even accept
    if (isPresent == false) {
        return;
    }

    // remove the annotation
    await removeSuggestion();

    editor = vscode.window.activeTextEditor;

    isBlocked = true;

    suggToApply = s;
    tmp = line.substr(startTyping, line.length).trimStart()
    if (s.startsWith(tmp)) {
        suggToApply = s.substr(tmp.length, s.length);
    }

    vscode.window.showTextDocument(editor.document, 1, false).then(e => {
        return e.edit(edit => {
            return edit.insert(editor.selection.active, suggToApply)
        })
    }).then(() => {
        isBlocked = false;
        // the recomndation has been accepted, therefore there isn't a recommendation anymore
        isPresent = false;
        // change the behavior of the shortcuts
        vscode.commands.executeCommand('setContext', 'myContext', isPresent);
    });
}

exports.acceptSuggestion = acceptSuggestion;

/**
 * create a decoration
 * @param {String} text the text of the recommendation
 * @param {vscode.Position} startPos the starting position of the recommendation
 * @param {String} suggestion the all recommedation
 * @returns {JSON} the decoration that will be applied
 */
function decorationFactory(text, startPos, suggestion, suggestionColor) {
    let scrollable = true;
    let decoration = {
        renderOptions: {
            after: {
                color: suggestionColor,
                contentText: text,
                fontWeight: 'normal',
                fontStyle: 'normal',
                textDecoration: `none;${scrollable ? '' : ' position: absolute;'}`,
            },
        }
    };
    // if there is a light theme applied, change the color of the suggestion
    if (theme.indexOf("Light") >= 0) {
        decoration.renderOptions.after.color = "#BEBEBE"
    }
    endPos = new vscode.Position(startPos.line, startPos.character + suggestion.length);
    decoration.range = new vscode.Range(startPos, endPos);
    return decoration;
}

function applyDecorations(userHasWrote) {
    // only single line annotations can be inserted, therefore for each line of the sugg, we need to perform an insert action
    decorations = []

    editor = vscode.window.activeTextEditor;
    // get the position of the cursor
    cursor = editor.selection.active;
    // for every line of suggestion create an annotation
    for (i = 0; i < suggestion.length; i++) {
        pos = undefined;
        if (i == 0) {
            pos = new vscode.Position(cursor.line, cursor.character - 1);
        } else {
            pos = new vscode.Position(cursor.line + i, 0);
        }

        decoration = decorationFactory(suggestion[i], pos, suggestion, suggestionColor);
        decorations.push(decoration);
    }
    if (userHasWrote) {
        // change the first decoration
        decorations[0].renderOptions.after.contentText = tmp;
    }
    // now that we have all the annotations, we can apply them
    editor.setDecorations(annotationDecoration, decorations);
}

/**
 * inject and display the suggestion to the user with the proper decoration
 * @param {String} url the url in which the NN is hosted
 * @param {Boolean} inCodeOrComment whether the cursor is in code or comment and cìshich type of comment (single line, multi line, javadoc)
 * @param {String} suggestionColor the color for the suggestion
 * @param {Integer} confidence the minimum confidence that a suggestion must have in order to be shown
 * @returns {void}
 */
async function showSuggestion(url, suggestionColor, inCodeOrComment, confidence, singleLineComment) {
    // if there is already a recommendation return, do not display another one
    if (isPresent == true) {
        return;
    }
    // get the suggestion from the NN
    [s, suggestion, lines, methodContent] = await neuralNetwork.getSuggestion(url, inCodeOrComment, confidence, singleLineComment);
    if (s == undefined) {
        handle.decreasePendingSuggestions();
        if (handle.getPendingSuggestions() == 1 || handle.getPendingSuggestions() == 0) {
            console.log("This was the last rec, setting noLongerNeeded = false")
            handle.setRecNoLongerNeeded(false)
        }
        return;
    }
    // if the recommendation is no longer needed
    if (handle.getRecNoLongerNeeded() == true) {
        console.log("The rec is no longer needed, decrese pending")
        // vscode.window.showInformationMessage('Recommendation no longer needed');
        // we have just blocked this pending suggestion, therefore decrease the number of pending suggestions
        handle.decreasePendingSuggestions();
        if (handle.getPendingSuggestions() == 1 || handle.getPendingSuggestions() == 0) {
            console.log("This was the last rec, setting noLongerNeeded = false")
            handle.setRecNoLongerNeeded(false)
        }
        return
    }



    // if the user has typed something in the mean while we need to skrink the recommendation
    // accordingly if the what the user wrote matches the suggestion
    if (handle.getUserInput() != "" && s.startsWith(handle.getUserInput())) {
        console.log("There is a match between what the user wrote and the rec")
        tmp = s.substring(handle.getUserInput().length)
        // the lines that we insert must not trigger 'onDidChangeTextDocument' therefore we block the function
        isBlocked = true;
        // insert new lines if needed
        await insertNewLines(lines);
        isBlocked = false;
        // apply the decorations to show the suggestion, but shrink the first line of suggestion,
        // because of the match
        applyDecorations(true, tmp)
        // set this position, so that we know what's the next char the user is supposed to write in the recommendation
        handle.setPositionInSuggestion(handle.getUserInput().length - 1)
    } else if (handle.getUserInput() != "") {
        console.log("The user has typed something else in the meanwhile that does not match the rec")
        return
    } else {
        // the lines that we insert must not trigger 'onDidChangeTextDocument' therefore we block the function
        isBlocked = true;
        // insert new lines if needed
        await insertNewLines(lines);
        isBlocked = false;
        // apply the decorations to show the suggestion
        applyDecorations(false, "")
    }

    // we are showing a suggestion
    isPresent = true;
    // the suggestion has been shown, therefore we do not have a pending suggestion
    handle.decreasePendingSuggestions();
    console.log("the suggestion has been shown")
    // change the behavior of the shortcuts
    vscode.commands.executeCommand('setContext', 'myContext', isPresent);
}

/**
 * this function reapplys the same suggestion again without the first char.
 * If the user types the suggestion, then the first char of the suggestion must dissappear because it has just been typed.
 * Example:
 * this is the suggestion: "comment suggetion stubbed"
 * if the user types: 'c'
 * then the suggestion must become: "omment suggestion stubbed"
 * if the user then types: 'o'
 * then the suggestion must become: "mment suggestion stubbed"
 * @param {vscode.Position} startPos the position in which the suggestion starts
 * @returns {void}
 */
const shrinkSuggestion = (startPos) => {
    tmp = decorations[0].renderOptions.after.contentText
    decorations[0].renderOptions.after.contentText = tmp.substring(1, tmp.length);
    endPos = new vscode.Position(startPos.line, startPos.character + tmp.length);
    decorations[0].range = new vscode.Range(startPos, endPos);
    editor.setDecorations(annotationDecoration, decorations);
}

exports.shrinkSuggestion = shrinkSuggestion;

/**
 * if we are deleting and there isn't a suggestion suggestion showing, then try to re-apply the suggestion if
 * the suggestion (partially) matches what's written in the current line.
 * Example:
 * This is the previous suggestion: this.max = 30;
 * The current line contains: this.income = 40;
 * If the user deletes the line like this: this.
 * then the suggestion should appear again, because the line prtially matches the suggestion.
 * @param {vscode.Position} startPos the position in which the suggestion starts
 * @param {String} line the content of the current line
 * @returns {boolean} true if metching has been found and a suggestion in shown, false otherwise
 */
const reApplySuggestion = (startPos, line) => {
    tmp = suggestion[0];
    if (tmp.startsWith(line)) {
        tmp = tmp.substr(line.length, tmp.length);
        // change the first decoration
        decorations[0].renderOptions.after.contentText = tmp;
        // recompute its endingPosition
        endPos = new vscode.Position(startPos.line, startPos.character + suggestion.length);
        decorations[0].range = new vscode.Range(startPos, endPos);
        // reapply the decorations
        editor.setDecorations(annotationDecoration, decorations);
        // there is now a recommnedation showing
        isPresent = true;
        // change the behavior of the shortcuts
        vscode.commands.executeCommand('setContext', 'myContext', isPresent);
        // we performed a match, return true
        return true;
    }
    // a matched was not found, return false
    return false;

}
exports.reApplySuggestion = reApplySuggestion;

/**
 * if there is a suggestion first remove it and then create another one
 * @param {String} url the url that hosts the NN 
 * @param {String} suggestionColor the color of the suggestion 
 * @param {Boolean} inCodeOrComment whether the cursor is in code or comment and cìshich type of comment (single line, multi line, javadoc)
 * @param {Integer} confidence the minimum confidence that a suggestion must have in order to be shown
 */
const removeAndShow = async (url, suggestionColor, inCodeOrComment, confidence, singleLineComment) => {
    await removeSuggestion();
    await showSuggestion(url, suggestionColor, inCodeOrComment, confidence, singleLineComment);
}

exports.removeAndShow = removeAndShow;

src/feedback.js
const fetch = require("node-fetch");


/**
 * Send the feedback for the given suggestion
 * @param {String} suggestion the suggestion returned by the NN
 * @param {String} methodContent the content of the method
 * @param {Boolean} isPositive whether the suggestion has been accepted or not 
 * @param {String} url the url that hosts the feedback service
 */
const sendFeedback = (suggestion, methodContent, isPositive, url, name) => {
    data = {
        "suggestion": suggestion,
        "methodContent": methodContent,
        "isPositive": isPositive
    }
    if (name != undefined && name.trim() != "") {
      data["name"] = name;
    }
    payload = {
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
        },
        method: "POST",
        body: JSON.stringify(data)
      }

      fetch(url, payload)


}

exports.sendFeedback = sendFeedback

src/extension.js
// --------------------------------------------
// this is the main file that is ran by Vs-Code
// --------------------------------------------

//To run the extenstion press 'fn'+'f5' --> VS code extension development'

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const parser = require('./parser')
const myEditor = require('./editor')
const handleKeyPresses = require('./handleKeyPresses')
const feedback = require('./feedback');
const fs = require('fs');

// this array contains the chars for which the plugin is going to be triggered
triggerChars = [" ", ".", ","];

// this counter keeps track of the previous position of the caret
oldLine = -1;
// this boolean tells you if the current line contains code (true) or a comment (false)
isCode = true;

// this variable holds the url to query for code completion
// this variable is set by defualt (see package.json)
// and can be changed in the settings
urlCode = ""
// this variable holds the url to query for comment completion
// this variable is set by defualt (see package.json)
// and can be changed in the settings
urlComment = ""

// this variable can be sent to the serve when asking for a recommnedation.
// this will enable the server to distiguish your requests from all the others
name = null;

// the color of the suggestion. Depanding of the vs-code theme the user is using, the default color (dark grey) could
// not be visible. This variable can be modified in the settings.
suggestionColor = "#505050"

// Specify the minimum confidence for the prediction returned by the Neural Network for code. 
// If confidence for the current prediction is below the specified number, the predictions will not be shown
confidenceCode = 90;

// Specify the minimum confidence for the prediction returned by the Neural Network for comments. 
// If confidence for the current prediction is below the specified number, the predictions will not be shown
confidenceComment = 90;

// if the cursor is inside a code or comment an which type of comment (single line, multi line, javadoc)
inCodeOrComment = false;

// the url for the server that stores the recommendations
urlFeedback = "http://gym.si.usi.ch:45002/feedback/"

// if true and a single line comment is dected on the same line of some code, move the comment to line above
singleLineComment = true;

/**
 * parse the array of json objects for urls defined in the settings
 * @param {Array[JSON]} arr the array of JSON objects representing the user settings for the given url
 * @returns {String || undefined}  the url field of the first JSON object with the enable parameter set to true,
 *                                  undefined if no such objects is found. 
 */
function parseJson(arr) {
    for (const o of arr) {
        // return the the first url enabled
        if (o.enable == true) {
            return o.url;
        }
    }
    return undefined;
}

/**
 * get the parameters set in the settings and set the variables accordingly
 * @returns {void}
 */
function getMyConfiguration() {
    // get all the settings
    configutation = vscode.workspace.getConfiguration('simon');

    // parse the settings for code completion
    urlCode = parseJson(configutation.urlCode);
    // set to defaulr url in case of undefined
    if (urlCode == undefined) {
        urlCode = "http://gym.si.usi.ch:45001/code/"
    }

    // parse the settings for comment completion
    urlComment = parseJson(configutation.urlComment);
    // set to default url in case of undefined
    if (urlComment == undefined) {
        urlComment = "http://gym.si.usi.ch:45001/comment/"
    }

    // get the chars that trigger code/comment completion
    triggerChars = configutation.triggerChars;

    // get the name parameter
    name = configutation.name;

    // get the suggestion color parameter
    suggestionColor = configutation.suggestionColor;

    // get the confidence parameter for the code NN
    confidenceCode = configutation.confidenceCode;

    // get the confidence parameter for the comment NN
    confidenceComment = configutation.confidenceComment;

    // get the feedbackUrl parameter
    urlFeedback = configutation.urlFeedback;

    // get the boolean to move single line comments to newlines
    singleLineComment = configutation.singleLineComment;

    console.log(urlCode, urlComment, triggerChars, name, suggestionColor, confidenceCode, confidenceComment, urlFeedback, singleLineComment)
}

/**
 * this function is called when the user types something
 * @returns {void}
 */
function handleChangeDocument() {

    if (myEditor.getIsBlocked() == true) {
        return;
    }


    editor = vscode.window.activeTextEditor

    // the extension is available only when working with java files
    if (editor.document.fileName.indexOf(".java") <= 0) {
        console.log("You are not working with a java file")
        return;
    }

    // get the current position of the caret
    caretPosition = editor.selection.active;
    // get the content of the current line
    documentLine = editor.document.lineAt(caretPosition)._text
    // fetch the char at the current position of the caret, this is the char that the user has just wrote
    charAtCaret = documentLine.charAt(caretPosition.character);
    // the length of the next line
    lineLength = editor.document.lineAt(caretPosition.line + 1)._text.trim().length;
    // if the user types one of these chars, set oldLine to -1 in order to recompute 
    // the kind of line the user is typing (comment or code)
    if (charAtCaret == "/" || charAtCaret == "*" || parser.getTypeOfLine() == 2) {
        console.log("Set oldLine = -1")
        oldLine = -1;
    }

    // the user has just pressed return and we were in a single line comment
    if ((lineLength == 0 && isNaN(charAtCaret.charCodeAt())) && ((inCodeOrComment == 1) || oldLine == -1)) {
        oldLine = -1;
        inCodeOrComment = 0;
        handleKeyPresses.handleKeyPresses(urlCode, triggerChars, suggestionColor, inCodeOrComment, confidenceCode, urlFeedback, singleLineComment, name);
        return;
    }

    // if since the previous char we didn't move this means that we don't have to try 
    // to understand again whether we are in a comment or not
    if (oldLine == caretPosition.line) {
        // if before we were in code then we call the code completion NN
        if (isCode) {
            console.log("we are in the same line and I call code")
            handleKeyPresses.handleKeyPresses(urlCode, triggerChars, suggestionColor, inCodeOrComment, confidenceCode, urlFeedback, singleLineComment, name);
        } else {
            // otherwise we call the comment completion NN
            console.log("we are in the same line and I call comment")
            handleKeyPresses.handleKeyPresses(urlComment, triggerChars, suggestionColor, inCodeOrComment, confidenceComment, urlFeedback, singleLineComment, name)
        }
    } else {
        // we changed line, now we have to understand whether we are in a comment or in code
        console.log("recomputing the line")
        oldLine = caretPosition.line;

        lt = parser.commentOrCode()
        inCodeOrComment = lt;
        if (lt == 0) {
            isCode = true;
            handleKeyPresses.handleKeyPresses(urlCode, triggerChars, suggestionColor, inCodeOrComment, confidenceCode, urlFeedback, singleLineComment, name);
        } else {
            isCode = false;
            handleKeyPresses.handleKeyPresses(urlComment, triggerChars, suggestionColor, inCodeOrComment, confidenceComment, urlFeedback, singleLineComment, name);
        }
    }  
    
}


// this is the main function in which you need to register the commands defined in package.json
async function activate(context) {

    // create the file to write the logs
    fs.writeFile("/Users/Shared/Omenetti_Matteo_Logs.txt", "", function(err) {
        if(err) {
            return console.log(err);
        }
    }); 

    getMyConfiguration();
    // if the user changes the setting this function is called
    context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(event => {
        getMyConfiguration();
    }));


    // this function is called when the user types something
    // (and also when the document is modified for example by a copy and paste)
    vscode.workspace.onDidChangeTextDocument((e) => {
        handleChangeDocument();
    });


    vscode.window.onDidChangeTextEditorSelection((e) => {
        // if the user moves the cursor with the mouse, then remove the suggestion
        if (e.kind == 2) {
            if (myEditor.getIsPresent() == true) {
                myEditor.removeSuggestion();
                //vscode.window.showInformationMessage('A suggestion has been rejected!');
                feedback.sendFeedback(s, myEditor.getMethodContent(), false, urlFeedback, name)
            }

        }
    })
    // this is the command to accept the suggestion
    let disposable = vscode.commands.registerCommand('simon.accept', function () {
        if (myEditor.getIsPresent() == false) {
            return
        }
        editor = vscode.window.activeTextEditor;
        if (editor == undefined) {
            return
        }
        // get the current position of the caret
        caretPosition = editor.selection.active;
        // get the content of the current line
        line = editor.document.lineAt(caretPosition)._text
        myEditor.acceptSuggestion(line, handleKeyPresses.getStartTyping());
        oldLine = -1;
        handleKeyPresses.setWasRec(false);
        // show a pup-up message for debugging purposes
        //vscode.window.showInformationMessage('A suggestion has been accepted!');
        feedback.sendFeedback(s, myEditor.getMethodContent(), true, urlFeedback, name)
    }, );

    context.subscriptions.push(disposable);

    // remove the suggestion manually with a shortcut
    disposable = vscode.commands.registerCommand('simon.delete', function () {
        if (myEditor.getIsPresent() == false) {
            return
        }
        myEditor.removeSuggestion();
        // show a pup-up message for debugging purposes
        // vscode.window.showInformationMessage('A suggestion has been rejected!');
        // send a negative feedback to the server
        feedback.sendFeedback(s, myEditor.getMethodContent(), false, urlFeedback, name)
    });
   

     // trigger the suggestion manually with a shortcut
     disposable = vscode.commands.registerCommand('simon.trigger', function () {

        editor = vscode.window.activeTextEditor
        // the extension is available only when working with java files
        if (editor.document.fileName.indexOf(".java") <= 0) {
            console.log("You are not working a java file")
            return;
        }
        myEditor.removeSuggestion()

        if (parser.commentOrCode() == 0) {
            myEditor.removeAndShow(urlCode, suggestionColor, inCodeOrComment, confidenceCode);
        } else {
            myEditor.removeAndShow(urlComment, suggestionColor, inCodeOrComment, confidenceComment);
        }
        // vscode.window.showInformationMessage('A suggestion has been triggered manually!');
    });
}


// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
    activate,
    deactivate
}